lib/fastlane/plugin/firebase_app_distribution/actions/firebase_app_distribution_action.rb

```diff
@@ -15,16 +16,74 @@ class FirebaseAppDistributionAction < Action
       FIREBASECMD_ACTION = "appdistribution:distribute".freeze
       BASE_URL = "https://firebaseappdistribution.googleapis.com"
       PATH = "/v1alpha/apps/"
+      MAX_POLLING_RETRIES = 60
+      POLLING_INTERVAL_S = 2
 
       extend Helper::FirebaseAppDistributionHelper
 
       def self.run(params)
         params.values # to validate all inputs before looking for the ipa/apk
-        validate_app!(params[:app])
+        platform = Actions.lane_context[Actions::SharedValues::PLATFORM_NAME]
+        binary_path = params[:ipa_path] || params[:apk_path]
+
+        if params[:app] # Set app_id if it is specified as a parameter
+          app_id = params[:app]
+        else
+          if platform == :ios
+            archive_path = Actions.lane_context[SharedValues::XCODEBUILD_ARCHIVE]
+            if archive_patharchive_path
+              app_id = findout_ios_app_id_from_archive(archive_path)
+            end
+          end
+          if app_id.nil?
+            app_id = UI.input("Your app's Firebase App ID. You can find the App ID in the Firebase console, on the General Settings page: ")
+          end
+        end
+
+        upload_token = get_upload_token(app_id, binary_path)
+        if upload_status(upload_token, app_id) == "SUCCESS"
+          UI.message("This APK/IPA has been uploaded before. Skipping upload step.")
+        else
+          MAX_POLLING_RETRIES.times do
+            status = upload_status(upload_token, app_id)
```

Comment:
Getting the status again here is unnecessary the first time through. Can you avoid it?

One way would be to save off the `status` during the first call to `upload_status`, and then move this call to the end of the loop.

Reply:
The change didn't actually remove the unnecessary call to `upload_status`, but now you can just move this call to the end of the loop (line 54).